File: 112. Path Sum/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  boolean dfs(TreeNode current, int amount, int targetSum){
    if (current == null){
      return false;
    }
    amount += current.val;
    if (current.left == null && current.right == null && amount == targetSum){
      return true;
    }
    return dfs(current.left, amount, targetSum) || dfs(current.right, amount, targetSum);
  }
  public boolean hasPathSum(TreeNode root, int targetSum) {
    int total = 0;
    return dfs(root, total, targetSum);
  }
}

File: 112. Path Sum/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  bool dfs(TreeNode* current, int amount, int targetSum){
    if (current == NULL){
      return false;
    }
    amount += current->val;
    if (current->left == NULL && current->right == NULL && amount == targetSum){
      return true;
    }
    return dfs(current->left, amount, targetSum) || dfs(current->right, amount, targetSum);
  }
public:
  bool hasPathSum(TreeNode* root, int targetSum) {
    int total = 0;
    return dfs(root, total, targetSum);
  }
};